python-lib/api_calling_utils.py

```diff
@@ -0,0 +1,306 @@
+try:
+    from requests.exceptions import RequestException
+    API_EXCEPTIONS = RequestException
+except ImportError:
+    pass
+try:
+    from boto3.exceptions import Boto3Error
+    from botocore.exceptions import BotoCoreError, ClientError
+    API_EXCEPTIONS = (Boto3Error, BotoCoreError, ClientError)
+except ImportError:
+    pass
+try:
+    from google.api_core.exceptions import GoogleAPICallError, RetryError
+    API_EXCEPTIONS = (GoogleAPICallError, RetryError)
+except ImportError:
+    pass
+
+# ==============================================================================
+# FUNCTION DEFINITION
+# ==============================================================================
+
+ApiColumnNameTuple = namedtuple("ApiColumnNameTuple", API_COLUMN_NAMES)
+
+
+def generate_unique(
+    name: AnyStr,
+    existing_names: List,
+    prefix: AnyStr = COLUMN_PREFIX
+) -> AnyStr:
+    """
+    Generate a unique name among existing ones by suffixing a number.
+    Can also add an optional prefix.
+    """
+    if prefix is not None:
+        new_name = prefix + "_" + name
+    else:
+        new_name = name
+    for j in range(1, 1000):
+        if new_name not in existing_names:
+            return new_name
+        new_name = name + "_{}".format(j)
+    raise Exception("Failed to generated a unique name")
+
+
+def initialize_api_column_names(
+    existing_names: List[AnyStr],
+    column_prefix: AnyStr = COLUMN_PREFIX
+) -> NamedTuple:
+    """
+    Helper function to the "api_parallelizer" main function.
+    Initializes a named tuple of column names from ApiColumnNameTuple,
+    adding a prefix and a number suffix to make them unique.
+    """
+    api_column_names = ApiColumnNameTuple(
+        *[generate_unique(k, existing_names, column_prefix)
+          for k in ApiColumnNameTuple._fields])
+    return api_column_names
+
+
+def safe_json_loads(
+    str_to_check: AnyStr,
+    error_handling: ErrorHandlingEnum = ErrorHandlingEnum.LOG
+) -> Dict:
+    """
+    Wrap json.loads with an additional parameter to handle errors:
+    - 'FAIL' to use json.loads, which fails on invalid data
+    - 'LOG' to try json.loads and return an empty dict if data is invalid
+    """
+    if error_handling == ErrorHandlingEnum.FAIL:
+        output = json.loads(str_to_check)
+    else:
+        try:
+            output = json.loads(str_to_check)
+        except (TypeError, ValueError):
+            logging.warning("Invalid JSON: '" + str(str_to_check) + "'")
+            output = {}
+    return output
+
+
+def fail_or_warn_row(
+    api_call_function: Callable,
+    api_column_names: NamedTuple,
+    row: Dict,
+    api_exceptions: Union[Exception, Tuple[Exception]] = API_EXCEPTIONS,
+    error_handling: ErrorHandlingEnum = ErrorHandlingEnum.LOG,
+    verbose: bool = False,
+    **api_call_function_kwargs
+) -> Dict:
+    """
+    Wraps a single-row API calling function to:
+    - ensure it has a 'row' parameter which is a dict (BATCH is *not* supported)
```

Comment:
Add ref to the batch function: instead of "BATCH is *not* supported", say: for batch of rows, use blabla below

Reply:
good point, will do

Reply:
done